R/__ddl_by_AC.R

```diff
@@ -0,0 +1,298 @@
+#' @param datanames character vector of names of data sets created; required for compatibility with `teal` apps
+#' @param join_keys `join_keys` object specifying relationships between data sets; defaults to `teal.data::join_keys()`
+#' @return A`reactive` expression returning a `tdata` object.
+#'
+#' #' @examples
+#' library(shiny)
+#' module <- input_template(
+#'   textInput("user", "username", placeholder = "who goes there?"),
+#'   passwordInput("pass", "password", placeholder = "friend or foe?"),
+#'   actionButton("submit", "get it"),
+#'   on_submit = function(input) {
+#'     example_data <- paste(input$user, input$pass, sep = " -- ")
+#'   },
+#'   mask = list(pass = "MASKED PASSWORD")
+#' )
+#' ui <- fluidPage(
+#'   tagList(
+#'     module$ui("id"),
+#'     verbatimTextOutput("value"),
+#'     verbatimTextOutput("code")
+#'   )
+#' )
+#' server <- function(input, output, session) {
+#'   tdata <- module$server("id")
+#'   output[["value"]] <- renderPrint({
+#'     req(tdata())
+#'     teal.code::get_var(tdata(), "example_data")
+#'   })
+#'   output[["code"]] <- renderPrint({
+#'     req(tdata())
+#'     cat(teal.code::get_code(tdata()), cat(sep = "\n"))
+#'   })
+#' }
+#' if (interactive()) shinyApp(ui, server)
+#'
+input_template <- function(..., on_submit, mask, datanames, join_keys) {
+  args <- list(...)
+  checkmate::assert_list(args, types = "shiny.tag")
+
+  args  <- as.list(substitute(list(...)))[-1L]
+  inputIds <- vapply(args, function(x) match.call(eval(x[[1L]]), x)[["inputId"]], character(1L))
+
+  checkmate::assert_true(
+    is.element("submit", inputIds),
+    .var.name = "A \"submit\" element is specified."
+  )
+
+  submit <- unlist(eval(args[[which(inputIds == "submit")]]))
+  submit_class <- submit[grep("class$", names(submit))]
+  checkmate::assert_true(
+    grepl("action-button", submit_class),
+    .var.name = "The \"submit\" element has class \"action-button\"."
+  )
+
+  # Wrap `inputIds` arguments in in `ns` calls.
+  args <- lapply(args, function(call) {
+    call <- match.call(eval(call[[1]]), call)
+    call <- as.list(call)
+    call[["inputId"]] <- call("ns", call[["inputId"]])
+    as.call(call)
+  })
+
+
+  ui <- function(id) {
+    ns <- NS(id)
+    div(
+      class = "connector-input",
+      lapply(args, eval, envir = environment())
+    )
+  }
+
+  checkmate::assert_character(datanames)
+  if (missing(mask)) mask <- list()
+  if (missing(join_keys)) join_keys <- teal.data::join_keys()
+  tracked_request <- with_substitution(on_submit, mask, join_keys)
+  server <- function(id) {
+    moduleServer(id, function(input, output, session) {
+      result <- eventReactive(input[["submit"]], {
+        inputs <- sapply(setdiff(inputIds, "submit"), function(x) input[[x]], simplify = FALSE)
+        tryCatch(
+          do.call(tracked_request, list(inputs)),
+          error = function(e) validate(need(FALSE, sprintf("Error: %s", e$message)))
+        )
+      })
+      result
+    })
```

Comment:
We need a possibility to specify this `server`. For example to control hierarchical inputs.
```suggestion
  server <- function(id) {
    moduleServer(id, function(input, output, session) {
      result <- eventReactive(input[["submit"]], {
        inputs <- sapply(setdiff(inputIds, "submit"), function(x) input[[x]], simplify = FALSE)
        tryCatch(
          do.call(tracked_request, list(inputs)),
          error = function(e) validate(need(FALSE, sprintf("Error: %s", e$message)))
        )
      })
      result
    })
```